2023/08/wtchangdm/day8.rs

```diff
@@ -0,0 +1,190 @@
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Copy)]
+enum Direction {
+    Left,
+    Right,
+}
+
+impl Direction {
+    fn from(ch: &char) -> Self {
+        match ch {
+            'L' => Self::Left,
+            'R' => Self::Right,
+            _ => panic!("come on it's impossible"),
+        }
+    }
+}
+
+struct Map {
+    cursor: usize,
+    directions: Vec<Direction>,
+    nodes: HashMap<String, [String; 2]>,
+}
+
+impl Map {
+    fn from(input: &[String]) -> Self {
+        let mut nodes: HashMap<String, [String; 2]> = HashMap::new();
+        let mut iter = input.iter().peekable();
+        let directions: Vec<Direction> = iter
+            .next()
+            .unwrap()
+            .chars()
+            .map(|ch| Direction::from(&ch))
+            .collect();
+
+        iter.next();
+
+        for line in iter {
+            let (current, adjacents) = line.split_once(" = ").unwrap();
+            let adjacents = adjacents.replace(['(', ')'], "");
+            let (left, right) = adjacents.split_once(", ").unwrap();
+
+            nodes
+                .entry(current.into())
+                .or_insert([left.to_string(), right.to_string()]);
+        }
+
+        Self {
+            directions,
+            nodes,
+            cursor: 0,
+        }
+    }
+
+    fn nodes(&self) -> &HashMap<String, [String; 2]> {
+        &self.nodes
+    }
```

Comment:
I feel like it's an unsuccessful abstraction. We have a `Map` type but still expose the entire `HashMap` implementation detail.

Not really a big deal for a puzzle solver :)

Reply:
True. Since these can just access the internals directly. Will revise along with other suggestions.